rust/main/submitter/src/chain_tx_adapter/chains/ethereum/adapter.rs

```diff
@@ -0,0 +1,177 @@
+
+        Ok(Self {
+            _conf: conf,
+            _raw_conf: raw_conf,
+            provider,
+            reorg_period,
+        })
+    }
+
+    async fn block_number_result_to_tx_hash(&self, block_number: Option<U64>) -> TransactionStatus {
+        let Some(block_number) = block_number else {
+            return TransactionStatus::PendingInclusion;
+        };
+        let block_number = block_number.as_u64();
+        match self
+            .provider
+            .get_finalized_block_number(&self.reorg_period)
+            .await
+        {
+            Ok(finalized_block) => {
+                if finalized_block as u64 >= block_number {
+                    TransactionStatus::Finalized
+                } else {
+                    TransactionStatus::Included
+                }
+            }
+            Err(err) => {
+                warn!(
+                    ?err,
+                    "Error checking block finality. Assuming tx is pending inclusion"
+                );
+                TransactionStatus::PendingInclusion
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl AdaptsChain for EthereumTxAdapter {
+    async fn estimate_gas_limit(
+        &self,
+        _payload: &FullPayload,
+    ) -> Result<Option<GasLimit>, SubmitterError> {
+        todo!()
+    }
+
+    async fn build_transactions(&self, payloads: &[FullPayload]) -> Vec<TxBuildingResult> {
+        use crate::chain_tx_adapter::chains::ethereum::transaction::TransactionFactory;
+
+        info!(?payloads, "building transactions for payloads");
+        let payloads_and_precursors = payloads
+            .iter()
+            .map(|payload| (EthereumTxPrecursor::from_payload(payload), payload))
+            .collect::<Vec<(EthereumTxPrecursor, &FullPayload)>>();
+
+        let mut transactions = Vec::new();
+        for (precursor, payload) in payloads_and_precursors.into_iter() {
+            let transaction = TransactionFactory::build(payload, precursor);
+            transactions.push(TxBuildingResult::new(
+                vec![payload.details.clone()],
+                Some(transaction),
+            ))
+        }
+
+        info!(?payloads, ?transactions, "built transactions for payloads");
+        transactions
+    }
+
+    async fn simulate_tx(&self, _tx: &Transaction) -> Result<bool, SubmitterError> {
+        todo!()
+    }
+
+    async fn estimate_tx(&self, _tx: &mut Transaction) -> std::result::Result<(), SubmitterError> {
+        todo!()
+    }
+
+    async fn submit(&self, tx: &mut Transaction) -> Result<(), SubmitterError> {
+        use crate::chain_tx_adapter::chains::ethereum::transaction::Precursor;
+
+        info!(?tx, "submitting transaction");
+
+        let precursor = tx.precursor();
+        let hash = self
+            .provider
+            .send(precursor.tx.clone(), precursor.function.clone())
+            .await?;
+
+        tx.tx_hashes.push(hash.into());
```

Comment:
realizing it's easy to forget to push the tx hash in future adapter implementations of `submit`. Should we change the signature of `submit` to return the tx hash on success, and push this to `tx_hashes` as part of the inclusion step?

Reply:
We update more data inside transaction in `SealevelTxAdapter`. It means that the return type will be more complicated than just hash. I think we can keep it as is.

Reply:
So, the transaction as parameter still will be modifiable inside the method, but we return hash and updated it outside of the method as well?